clang/docs/checkedc/Bounds-Checking-for-LValue-Expressions.md

```diff
@@ -0,0 +1,268 @@
+checking variable bounds:
+
+- ObservedBounds: maps a variable declaration (`VarDecl *`) to the current
+  bounds (`BoundsExpr *`) that the compiler has inferred for the variable.
+- EquivExprs: set of sets of expressions. If two expressions `e1` and `e2`
+  are in the same set in `EquivExprs`, then `e1` and `e2` produce the same
+  value.
+
+### TraverseCFG
+
+This method traverses each block in the body of a function. Before traversing
+the function body, each function parameter `p` with declared bounds is added
+to ObservedBounds. For example:
+
+```
+void f(_Array_ptr<int> a : count(len), int len) {
+	// ObservedBounds = { a => bounds(a, a + len) }.
+
+	...Body of f
+}
+```
+
+### GetIncomingBlockState
+
+Before traversing a CFG block `B`, this method sets `ObservedBounds` and
+`EquivExprs` to the intersection of the `ObservedBounds_i` and `EquivExprs_i`
+from each predecessor block `B_i` of `B`. This ensures, that, before checking
+a statement `S` within block `B`, `ObservedBounds` contains only variables that
+are in scope at `S`. For example:
+
+```
+void f(int flag) {
+	// Block B1.
+	// At the end of this block, ObservedBounds contains bounds for a.
+	_Array_ptr<int> a : count(1) = 0;
+
+	if (flag) {
+		// Block B2. Predecessors: B1.
+		// At the beginning of this block: ObservedBounds contains a.
+		// At the end of this block: ObservedBounds contains a and b.
+		_Array_ptr<int> b : count(2) = 0;
+	}
+
+	// Block B3. Predecessors: B1, B2.
+	// B1's ObservedBounds contains a. B2's observed bounds contains a and b.
+	// At the beginning of this block: ObservedBounds contains a.
+	// At the end of this block: ObservedBounds contains a and c.
+	_Array_ptr<int> c : count(3) = 0;
+}
+```
+
+### UpdateCtxWithWidenedBounds
+
+Before checking a CFG block `B`, this method updates `ObservedBounds` to map
+each variable that has widened bounds in `B` to its widened bounds. For
+example:
+
+```
+void f(_Nt_array_ptr<char> p : count(1)) {
+	if (*(p + 1)) {
+		// CFG block B.
+		// Within this block, the bounds of p are widened by 1.
+
+		// p[1] is within the (widened) observed bounds of (p, p + 2).
+		char c = p[1];
+	}
+}
+```
+
+`p` initially has observed bounds of `bounds(p, p + 1)` (the declared bounds
+of `p`). In block `B`, `p` has widened bounds of `bounds(p, p + 2)`. Before
+checking block `B`, `UpdateCtxWithWidenedBounds` updates `ObservedBounds` so
+that `ObservedBounds[p] = bounds(p, p + 2)`.
+
+### GetDeclaredBounds
+
+Before checking a statement `S` within a CFG block `B`, this method updates
+`ObservedBounds` so that, for each variable `x` declared in `S` that has
+declared bounds `D`, `ObservedBounds[x] = D`. Before checking `S`, the observed
+bounds for each variable `v` that is in scope at `S` will be either:
+
+1. The widened bounds for `v`, if `v` has widened bounds within the current CFG
+   block `B`, or:
+2. The declared bounds for `v`, if `v` has declared bounds, or:
```

Comment:
`v` here could be a variable that was declared prior to `S`. For example:

```
void f() {
  _Array_ptr<int> a : count(1) = 0;

  // Statement S
  // Before checking S, the ObservedBounds context is { a => bounds(a, a + 1), b => bounds(b, b + 2) }
  _Array_ptr<int> b : count(2) = 0;
}
```

I moved this description of the initial `ObservedBounds` to the end of the section describing ResetKilledBounds, since at that point all the methods that are involved in setting the initial `ObservedBounds` have been described.

Reply:
Ok, thanks for the clarification!